crates/biome_js_analyze/src/semantic_analyzers/nursery/no_then_property.rs

```diff
@@ -0,0 +1,423 @@
+use crate::semantic_services::Semantic;
+use biome_analyze::{context::RuleContext, declare_rule, Rule, RuleDiagnostic};
+use biome_js_syntax::{
+    AnyJsArrayElement, AnyJsAssignment, AnyJsAssignmentPattern, AnyJsCallArgument,
+    AnyJsClassMemberName, AnyJsDeclarationClause, AnyJsExportClause, AnyJsExportNamedSpecifier,
+    AnyJsExpression, AnyJsObjectMemberName, AnyJsTemplateElement, JsMethodObjectMember,
+};
+use biome_js_syntax::{
+    AnyJsClassMember, AnyJsObjectMember, JsAssignmentExpression, JsCallExpression,
+    JsComputedMemberName, JsExport,
+};
+use biome_rowan::{declare_node_union, AstNode, AstSeparatedList, TextRange};
+
+declare_rule! {
+    /// Disallow `then` property
+    ///
+    /// When combining objects with a `then`` method (thenable objects) with await expressions or dynamic imports, caution is necessary.
+    /// These syntaxes interpret the object's then method as intended for the resolution or rejection of a promise, which can lead to unexpected behavior or errors.
+    ///
+    /// Source: https://github.com/sindresorhus/eslint-plugin-unicorn/blob/main/docs/rules/no-thenable.md
+    ///
+    /// ## Examples
+    ///
+    /// ### Invalid
+    ///
+    /// ```js,expect_diagnostic
+    /// export {then};
+    /// ```
+    ///
+    /// ```js,expect_diagnostic
+    /// const foo = {
+    ///     then() {}
+    /// };
+    /// ```
+    ///
+    /// ```js,expect_diagnostic
+    /// const foo = {
+    ///     get then() {}
+    /// };
+    /// ```
+    ///
+    /// ```js,expect_diagnostic
+    /// const foo = {
+    ///    get then() {}
+    /// };
+    /// ```
+    ///
+    /// ```js,expect_diagnostic
+    /// foo.then = function () {}
+    /// ```
+    ///
+    /// ```js,expect_diagnostic
+    /// class Foo {
+    ///     then() {}
+    /// }
+    /// ```
+    ///
+    /// ```js,expect_diagnostic
+    /// class Foo {
+    ///     static then() {}
+    /// }
+    /// ```
+    ///
+    /// ## Valid
+    ///
+    /// ```js
+    /// export {then as success};
+    /// ```
+    ///
+    /// ```js
+    /// const foo = {
+    ///     success() {}
+    /// };
+    /// ```
+    ///
+    /// ```js
+    /// class Foo {
+    ///     success() {}
+    /// }
+    /// ```
+    ///
+    /// ```js
+    /// const foo = bar.then;
+    /// ```
+    ///
+    pub(crate) NoThenProperty {
+        version: "next",
+        name: "noThenProperty",
+        recommended: false,
```

Comment:
@Conaclos I think we can recommend this rule following original eslint-plugin-unicorn, what do you think?

Reply:
I agree